### DISCUSSION TITLE
Rule change behavior with dependency of time of day

### DISCUSSION THREAD

**Discussion by igrowing** [arendst/Tasmota#17929] [Q&A] [answered]
Trying to figure out how to write a rule for standard [release of tasmota v.12.3.1](https://ota.tasmota.com/tasmota/release/) (tasmota.bin or tasmota-lite.bin) for ESP8266 which on button press turns Power1 for 2 minutes always, but the power2 should turn on only if the button is pressed in certain time interval of the day.

The rule below does not work as it contains `if` statement.

```
Rule1
  ON button1#state DO Backlog Power1 %value%; if [time("HHMM")>1700 or time("HHMM")<1710] do power2 %value% endon; RuleTimer1 10 ENDON
  ON Rules#Timer=1 DO Backlog Power1 off; power2 off ENDON
```

Advice would be appreciated.

**Comment by pkkrusty** [comment, marked as answer]
Easiest would be assigning var1 a 1 or 0 based on time of day, then assigning Power2 to that value. I.e. `rule1 on time#minute=1020 do var1 1 endon on time#minute=1030 do var1 0 endon on button1#state do backlog power1 1; power2 %var1%; ruletimer1 600 endon on rules#timer=1 do backlog power1 0; power2 0 endon`

**Reply by sfromis** [reply]
It is not only a question of `if` that the pseudocode is not rules, as inventions like this is unrelated to how rules work:
`[time("HHMM")>1700 or time("HHMM")<1710]`

(Rules does have `if` you build to include this optional features, but things can also be done without this, as long as you start by forgetting about usual concepts of programming)

**Reply by igrowing** [reply]
@pkkrusty Thank you! It took me time to realize that `1020` is not HHMM but total minutes from the beginning of the day. Ended up with the following:
```
var1 0
rule1 
  on time#minute=710 do var1 1 endon 
  on time#minute=720 do var1 0 endon 
  on switch3#state!=0 do backlog power1 1; power2 %var1% endon 
  ON power1#state DO RuleTimer1 10 ENDON
  on rules#timer=1 do backlog power1 0; power2 0 endon
```
The only drawback to this is: if it was power outage in between tiggering minutes then the state of var1 is default. Tried to use `<` and `>` for comparison of the trigger but this was bad idea.

**Reply by pkkrusty** [reply]
You could use < or > and enable one-shot (rule1 5) so that it only evaluates the trigger once if conditions stay the same. That way the console isn't flooded with messages. Then you get a continual testing for correct state of var1.

Although now that I think about it, that wouldn't work since both triggers would happen outside the window. 

You could `on time#initialized do event checktime %time% endon on event#checktime>710 do add2 1 endon on event#checktime<720 do add2 1 endon on var2=2 do backlog power1 1; power2 1 endon` and add your other logic.

**Reply by barbudor** [reply]
You can reach something similar to IF using BREAK
```
on time#minute>720 do var1 0 break on time#minute>=710 do var1 1 break on time#minute do var1 0
```
The order is of importance
BREAK acts as kind of a ELSE when the same trigger (time#minute) is used

Note that it will generate a log every minute because every minute one of the conditions will be met
It may be possible to limit that placing these into a separate ruleset and enabling 'once' mode

**Reply by igrowing** [reply]
Thank you @pkkrusty and @barbudor . Both solutions (one-shot and break) work.

**Reply by pkkrusty** [reply]
I like barbudor's suggestion. Simpler than mine.

**Reply by sfromis** [reply]
Beware that the method with `break` only works when the triggers are for the same event, typically same trigger name.
